core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -438,12 +439,27 @@ class BlockManagerMasterEndpoint(
     if (blockLocations.containsKey(blockId)) blockLocations.get(blockId).toSeq else Seq.empty
   }
 
-  private def getLocationsAndStatus(blockId: BlockId): Option[BlockLocationsAndStatus] = {
+  private def getLocationsAndStatus(
+      blockId: BlockId, requesterHost: String): Option[BlockLocationsAndStatus] = {
     val locations = Option(blockLocations.get(blockId)).map(_.toSeq).getOrElse(Seq.empty)
     val status = locations.headOption.flatMap { bmId => blockManagerInfo(bmId).getStatus(blockId) }
 
     if (locations.nonEmpty && status.isDefined) {
-      Some(BlockLocationsAndStatus(locations, status.get))
+      val bmIdToLocalDirs = if (status.get.storageLevel.useDisk) {
+        locations
+          .find(_.host == requesterHost)
```

Comment:
if its been cached by multiple executors on the same host, this will only return one of them.  I guess that is OK?  there isn't really an important scenario where it would become unavailable from one executor but still available on another.

Reply:
Yes that was my idea. First it is mostly only cached at one executor and even more rare to be cached/replicated on multiple executors but on the same host. So to give back only one of `localDirs` should be fine here.